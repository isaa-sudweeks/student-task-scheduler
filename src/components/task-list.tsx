"use client";

import React, { useState, useEffect, useRef } from "react";
import { toast } from "react-hot-toast";
import { api } from "@/server/api/react";
import type { RouterOutputs } from "@/server/api/root";

import { TaskListSkeleton } from "./task-list-skeleton";
import { TaskFilterTabs } from "./task-filter-tabs";
import {
  DndContext,
  closestCenter,
  type DragEndEvent,
} from "@dnd-kit/core";
import {
  SortableContext,
  useSortable,
  arrayMove,
} from "@dnd-kit/sortable";
import { CSS } from "@dnd-kit/utilities";
import { useVirtualizer } from "@tanstack/react-virtual";
import { Calendar, Tag, GripVertical } from "lucide-react";
import { TaskModal } from "@/components/task-modal";
import { StatusDropdown, type TaskStatus } from "@/components/status-dropdown";
import { Button } from "@/components/ui/button";

type Task = RouterOutputs["task"]["list"][number];
type Priority = "LOW" | "MEDIUM" | "HIGH";

export function TaskList() {
  const [filter, setFilter] = useState<"all" | "overdue" | "today" | "archive">("all");
  const [subject, setSubject] = useState<string | null>(null);
  const [priority, setPriority] = useState<Priority | null>(null);
  const [query, setQuery] = useState("");
  const utils = api.useUtils();

  const queryInput = React.useMemo(() => {
    const tzOffsetMinutes = new Date().getTimezoneOffset();
    const now = new Date();
    const startLocal = new Date(now);
    startLocal.setHours(0, 0, 0, 0);
    const endLocal = new Date(now);
    endLocal.setHours(23, 59, 59, 999);
    // Pass both explicit bounds and offset (offset kept for backward compatibility/tests)
    return {
      filter,
      subject: subject ?? undefined,
      priority: priority ?? undefined,
      tzOffsetMinutes,
      todayStart: startLocal,
      todayEnd: endLocal,
    } as const;
  }, [filter, subject, priority]);

  const tasks = api.task.list.useQuery(queryInput);
  // Query archived count for header stats
  const archivedQueryInput = React.useMemo(
    () => ({
      filter: "archive" as const,
      tzOffsetMinutes: queryInput.tzOffsetMinutes,
      todayStart: queryInput.todayStart,
      todayEnd: queryInput.todayEnd,
    }),
    [queryInput]
  );
  const archived = api.task.list.useQuery(archivedQueryInput);
  // Keep a stable snapshot of the fetched tasks for the current filter
  const [taskDataSnapshot, setTaskDataSnapshot] = useState<typeof tasks.data>();
  const prevFilterRef = React.useRef(filter);
  useEffect(() => {
    if (!tasks.data) return;
    // On first data load or when filter changes, capture snapshot
    if (taskDataSnapshot === undefined || prevFilterRef.current !== filter) {
      setTaskDataSnapshot(tasks.data);
      prevFilterRef.current = filter;
      return;
    }
    // If the set of task IDs has changed, refresh snapshot (create/delete)
    const snapIds = new Set((taskDataSnapshot as Task[]).map((t: Task) => t.id));
    const dataIds = new Set((tasks.data as Task[]).map((t: Task) => t.id));
    const idsDiffer =
      snapIds.size !== dataIds.size ||
      [...dataIds].some((id) => !snapIds.has(id));
    if (idsDiffer) {
      setTaskDataSnapshot(tasks.data);
      return;
    }
    // If content changed for existing IDs, refresh snapshot when updatedAt differs
    const snapById = new Map((taskDataSnapshot as Task[]).map((t: Task) => [t.id, t]));
    const contentChanged = (tasks.data as Task[]).some((t: Task) => {
      const prev = snapById.get(t.id);
      if (!(t && prev)) return false;
      // Detect edits via updatedAt change only to avoid noisy re-snapshots in tests
      if (t.updatedAt && prev.updatedAt) {
        const nextTs = new Date(t.updatedAt).getTime();
        const prevTs = new Date(prev.updatedAt).getTime();
        return nextTs !== prevTs;
      }
      return false;
    });
    if (contentChanged) {
      setTaskDataSnapshot(tasks.data);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [filter, tasks.data, taskDataSnapshot]);
  const [items, setItems] = useState<string[]>([]);

  useEffect(() => {
    if (tasks.data) {
      setItems((tasks.data as Task[]).map((t: Task) => t.id));
    }
  }, [tasks.data]);

  useEffect(() => {
    setSelected((prev) => new Set([...prev].filter((id) => items.includes(id))));
  }, [items]);

  const totalTasks = tasks.data?.length ?? 0;
  const archivedCount = archived.data?.length ?? 0;

  const setDue = api.task.setDueDate.useMutation({
    onSuccess: async () => utils.task.list.invalidate(),
    onError: (e) => toast.error(e.message || "Failed to set due date"),
  });

  // Inline rename/delete removed in minimalist UI; handled in modal

  const setStatus = api.task.setStatus.useMutation({
    onSuccess: async () => utils.task.list.invalidate(),
    onError: (e) => toast.error(e.message || "Failed to update status"),
  });

  const reorder = api.task.reorder.useMutation({
    onSuccess: async () => utils.task.list.invalidate(),
    onError: (e) => toast.error(e.message || "Failed to reorder tasks"),
  });

  const handleDragEnd = (event: DragEndEvent) => {
    const { active, over } = event;
    if (!over || active.id === over.id) return;
    setItems((prev) => {
      const oldIndex = prev.indexOf(active.id as string);
      const newIndex = prev.indexOf(over.id as string);
      const newItems = arrayMove(prev, oldIndex, newIndex);
      const snapshot = prev;
      reorder.mutate(
        { ids: newItems },
        {
          onError: () => setItems(snapshot),
        }
      );
      return newItems;
    });
  };

  const taskData = taskDataSnapshot ?? tasks.data;
  const orderedTasks = React.useMemo(() => {
    const list = taskData ?? [];
    if (items.length === 0) return list;
    const map = new Map((list as Task[]).map((t: Task) => [t.id, t]));
    return items.map((id) => map.get(id)).filter(Boolean) as Task[];
  }, [taskData, items]);

  const filteredOrderedTasks = React.useMemo(
    () =>
      orderedTasks.filter(
        (t) =>
          t.title.toLowerCase().includes(query.toLowerCase()) &&
          (!subject || t.subject === subject) &&
          (!priority || t.priority === priority)
      ),
    [orderedTasks, query, subject, priority]
  );
  // Compute the visible ids in the current order; feed to SortableContext
  const visibleIds = React.useMemo(
    () => filteredOrderedTasks.map((t) => t.id),
    [filteredOrderedTasks]
  );

  const [editingTask, setEditingTask] = useState<Task | null>(null);
  const [selected, setSelected] = useState<Set<string>>(new Set());

  const toggleSelected = (id: string) => {
    setSelected((prev) => {
      const next = new Set(prev);
      if (next.has(id)) next.delete(id);
      else next.add(id);
      return next;
    });
  };

  const bulkUpdate = api.task.bulkUpdate.useMutation({
    onSuccess: async () => {
      setSelected(new Set());
      await utils.task.list.invalidate();
    },
    onError: (e) => toast.error(e.message || "Failed to update tasks"),
  });

  const bulkDelete = api.task.bulkDelete.useMutation({
    onSuccess: async () => {
      setSelected(new Set());
      await utils.task.list.invalidate();
    },
    onError: (e) => toast.error(e.message || "Failed to delete tasks"),
  });

  const parentRef = useRef<HTMLDivElement>(null);
  const rowVirtualizer = useVirtualizer({
    count: filteredOrderedTasks.length,
    getScrollElement: () => parentRef.current,
    estimateSize: () => 64,
  });
  const useVirtual = filteredOrderedTasks.length >= 20;

  const TaskItem = ({
    t,
    virtualStyle,
  }: {
    t: Task;
    virtualStyle?: React.CSSProperties;
  }) => {
    const { attributes, listeners, setNodeRef, transform, transition } = useSortable({ id: t.id });
    const style: React.CSSProperties = {
      ...virtualStyle,
      ...(transform ? { transform: CSS.Transform.toString(transform) } : {}),
      ...(transition ? { transition } : {}),
    };
    const overdue = t.dueAt ? new Date(t.dueAt) < new Date() : false;
    const done = t.status === "DONE";
    const priority: Priority = t.priority ?? "MEDIUM";
    const priorityStyles: Record<Priority, string> = {
      HIGH: "bg-red-100 text-red-700 ring-red-200 dark:bg-red-900/30 dark:text-red-200 dark:ring-red-800",
      MEDIUM: "bg-blue-100 text-blue-700 ring-blue-200 dark:bg-blue-900/30 dark:text-blue-200 dark:ring-blue-800",
      LOW: "bg-slate-100 text-slate-700 ring-slate-200 dark:bg-slate-800 dark:text-slate-200 dark:ring-slate-700",
    };
    return (
      <li
        ref={setNodeRef}
        style={style}
        {...attributes}
        key={t.id}
        className={`flex items-center justify-between rounded border px-3 py-2 transition-colors hover:bg-black/5 dark:hover:bg-white/5 ${
          overdue
            ? "border-red-500 bg-red-50 text-red-800 dark:bg-red-950 dark:text-red-200"
            : ""
        }`}
        onClick={() => setEditingTask(t)}
      >
        <div className="flex items-start gap-3 flex-1">
          <input
            type="checkbox"
            aria-label="Select task"
            className="mt-1"
            checked={selected.has(t.id)}
            onChange={(e) => {
              e.stopPropagation();
              toggleSelected(t.id);
            }}
            onClick={(e) => e.stopPropagation()}
          />
          <button
            type="button"
            aria-label="Drag to reorder"
            className="mt-0.5 -ml-1 cursor-grab touch-none select-none rounded p-1 text-slate-400 hover:text-slate-600 active:cursor-grabbing dark:text-slate-500 dark:hover:text-slate-300"
            {...attributes}
            {...listeners}
            onClick={(e) => e.stopPropagation()}
          >
            <GripVertical className="h-4 w-4" />
          </button>
          <StatusDropdown
            className="mt-0.5"
            value={t.status as TaskStatus}
            onChange={(next) =>
              setStatus.mutate({ id: t.id, status: next })
            }
          />
          <div className="flex flex-col gap-1 flex-1">
            <div className="flex flex-wrap items-center gap-2">
              <span className={`font-medium ${done ? "line-through opacity-60" : ""}`}>
                {t.title}
              </span>
              <span
                className={`inline-flex items-center gap-1 rounded-full px-2 py-0.5 text-[10px] font-medium ring-1 ${priorityStyles[priority]}`}
              >
                {priority.charAt(0) + priority.slice(1).toLowerCase()}
              </span>
              {t.subject && (
                <span className="inline-flex items-center gap-1 rounded-full bg-slate-100 px-2 py-0.5 text-[10px] font-medium text-slate-700 ring-1 ring-slate-200 dark:bg-slate-800 dark:text-slate-200 dark:ring-slate-700">
                  <Tag className="h-3 w-3" /> {t.subject}
                </span>
              )}
              {t.dueAt && (
                <span className={`inline-flex items-center gap-1 rounded-full px-2 py-0.5 text-[10px] font-medium ring-1 ${
                  overdue
                    ? "bg-red-100 text-red-700 ring-red-200 dark:bg-red-900/30 dark:text-red-200 dark:ring-red-800"
                    : "bg-amber-100 text-amber-700 ring-amber-200 dark:bg-amber-900/30 dark:text-amber-200 dark:ring-amber-800"
                }`}>
                  <Calendar className="h-3 w-3" />
                  {new Date(t.dueAt!).toLocaleString()}
                </span>
              )}
            </div>
          </div>
        </div>
        {/* Right side kept minimal on list */}
      </li>
    );
  };

  return (
    <div className="space-y-3">
      <input
        type="text"
        value={query}
        onChange={(e) => setQuery(e.currentTarget.value)}
        placeholder="Search tasks..."
        className="w-full bg-transparent border-0 px-0 py-1 outline-none placeholder:text-muted-foreground"
      />
      <div className="flex items-center justify-between">
        <TaskFilterTabs
          value={filter}
          onChange={setFilter}
          subject={subject}
          onSubjectChange={setSubject}
          priority={priority}
          onPriorityChange={setPriority}
        />
        <p className="text-xs text-gray-500 dark:text-gray-400">
          {archivedCount} archived
        </p>
      </div>
      <DndContext collisionDetection={closestCenter} onDragEnd={handleDragEnd}>
        <SortableContext items={visibleIds}>
          {useVirtual ? (
            <div ref={parentRef} className="overflow-auto max-h-[600px]">
              <div
                style={{
                  height: rowVirtualizer!.getTotalSize(),
                  position: "relative",
                }}
              >
                {rowVirtualizer!.getVirtualItems().map((virtualRow) => {
                  const t = filteredOrderedTasks[virtualRow.index];
                  return (
                    <TaskItem
                      key={t.id}
                      t={t}
                      virtualStyle={{
                        position: "absolute",
                        top: 0,
                        left: 0,
                        width: "100%",
                        transform: `translateY(${virtualRow.start}px)`,
                        height: `${virtualRow.size}px`,
                      }}
                    />
                  );
                })}
              </div>
            </div>
          ) : (
            <ul className="space-y-2">
              {filteredOrderedTasks.map((t) => (
                <TaskItem key={t.id} t={t} />
              ))}
              {tasks.isLoading && <TaskListSkeleton />}
              {!tasks.isLoading && filteredOrderedTasks.length === 0 && (
                <li className="opacity-60">No tasks.</li>
              )}
            </ul>
          )}
        </SortableContext>
        {useVirtual && !tasks.isLoading && filteredOrderedTasks.length === 0 && (
          <div className="opacity-60">No tasks.</div>
        )}
        {useVirtual && tasks.isLoading && <TaskListSkeleton />}
      </DndContext>

      {tasks.error && (
        <p role="alert" className="text-red-500">
          {tasks.error.message}
        </p>
      )}
      {setDue.error && (
        <p role="alert" className="text-red-500">{setDue.error.message}</p>
      )}

      {selected.size > 0 && (
        <div data-testid="bulk-actions" className="flex gap-2">
          <Button
            onClick={() =>
              bulkUpdate.mutate({
                ids: Array.from(selected),
                status: "DONE",
              })
            }
          >
            Mark done
          </Button>
          <Button
            variant="danger"
            onClick={() =>
              bulkDelete.mutate({ ids: Array.from(selected) })
            }
          >
            Delete
          </Button>
        </div>
      )}

      <TaskModal
        open={!!editingTask}
        mode="edit"
        onClose={() => setEditingTask(null)}
        task={editingTask ?? undefined}
      />
    </div>
  );
}
