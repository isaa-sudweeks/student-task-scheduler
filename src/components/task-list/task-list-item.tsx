import React from "react";
import { useSortable } from "@dnd-kit/sortable";
import { CSS } from "@dnd-kit/utilities";
import { motion } from "framer-motion";
import { Calendar, Tag, Clock, GripVertical, MoreVertical } from "lucide-react";
import type Fuse from "fuse.js";

import { StatusDropdown, type TaskStatus } from "@/components/status-dropdown";
import type { Task } from "./use-task-list-query";

function highlightMatches(text: string, indices: readonly [number, number][]) {
  let last = 0;
  const res: React.ReactNode[] = [];
  indices.forEach(([start, end], i) => {
    if (start > last) res.push(text.slice(last, start));
    res.push(<mark key={i}>{text.slice(start, end + 1)}</mark>);
    last = end + 1;
  });
  if (last < text.length) res.push(text.slice(last));
  return res;
}

export const TaskCard = React.forwardRef<
  HTMLLIElement,
  {
    t: Task;
    style?: React.CSSProperties;
    listeners?: any;
    attributes?: any;
    isDragging?: boolean;
    onClick?: () => void;
    depth?: number;
    match?: readonly Fuse.FuseResultMatch[];
    onStatusChange: (next: TaskStatus) => void;
  }
>(({ t, style, listeners, attributes, isDragging, onClick, depth = 0, match, onStatusChange }, ref) => {
  const overdue = t.dueAt ? new Date(t.dueAt) < new Date() : false;
  const done = t.status === "DONE";
  const dueDate = t.dueAt ? new Date(t.dueAt) : null;
  let dueClass = "text-neutral-500";
  if (dueDate) {
    const now = new Date();
    const start = new Date(now);
    start.setHours(0, 0, 0, 0);
    const end = new Date(now);
    end.setHours(23, 59, 59, 999);
    if (dueDate < start) dueClass = "text-red-600";
    else if (dueDate <= end) dueClass = "text-amber-600";
  }
  const matchTitle = match?.find((m) => m.key === "title");
  const titleNode =
    matchTitle && matchTitle.indices
      ? highlightMatches(t.title, matchTitle.indices as any)
      : t.title;

  return (
    <motion.li
      ref={ref}
      style={{ ...style, marginLeft: depth * 16 }}
      layout
      initial={{ opacity: 0, scale: 0.98 }}
      animate={{ opacity: 1, scale: 1 }}
      exit={{ opacity: 0, scale: 0.98 }}
      transition={{ duration: 0.2, ease: "easeInOut" }}
      whileHover={{ scale: 1.02, transition: { duration: 0.15, ease: "easeInOut" } }}
      whileTap={{ scale: 0.98, transition: { duration: 0.15, ease: "easeInOut" } }}
      className={`group flex items-center justify-between rounded-md border bg-white p-3 transition-colors hover:bg-neutral-50 ${
        overdue
          ? "border-red-500 bg-red-50 text-red-800 dark:bg-red-950 dark:text-red-200"
          : ""
      } ${isDragging ? "shadow-sm ring-1 ring-neutral-200 translate-y-[-1px]" : ""}`}
      onClick={onClick}
    >
      <div className="flex items-start gap-3 flex-1">
        <button
          type="button"
          aria-label="Drag to reorder"
          className="cursor-grab text-neutral-400 hover:text-neutral-700 active:cursor-grabbing"
          {...attributes}
          {...listeners}
          onClick={(e) => e.stopPropagation()}
        >
          <GripVertical className="h-4 w-4" />
        </button>
        <div className="flex flex-col flex-1">
          <span className={`font-medium ${done ? "line-through opacity-60" : ""}`}>{titleNode}</span>
          {t.course?.title && (
            <span className="text-sm text-neutral-500">{t.course.title}</span>
          )}
          <div className="mt-1 flex flex-wrap items-center gap-4 text-xs text-neutral-500">
            {t.dueAt && (
              <span data-testid="due-date" className={`flex items-center gap-1 ${dueClass}`}>
                <Calendar className="h-4 w-4 text-neutral-400" />
                {dueDate!.toLocaleDateString()}
              </span>
            )}
            {t.subject && (
              <span className="flex items-center gap-1">
                <Tag className="h-4 w-4 text-neutral-400" />
                {t.subject}
              </span>
            )}
            {t.effortMinutes && (
              <span className="flex items-center gap-1">
                <Clock className="h-4 w-4 text-neutral-400" />
                {t.effortMinutes}m
              </span>
            )}
          </div>
        </div>
      </div>
      <div className="flex items-center gap-2">
        <StatusDropdown
          value={t.status as TaskStatus}
          onChange={onStatusChange}
        />
        <button
          type="button"
          aria-label="More actions"
          className="p-1 text-neutral-400 hover:text-neutral-700"
          onClick={(e) => {
            e.stopPropagation();
            onClick?.();
          }}
        >
          <MoreVertical className="h-4 w-4" />
        </button>
      </div>
    </motion.li>
  );
});
TaskCard.displayName = "TaskCard";

export const TaskListItem = React.forwardRef<
  HTMLLIElement,
  {
    t: Task;
    depth: number;
    virtualStyle?: React.CSSProperties;
    match?: readonly Fuse.FuseResultMatch[];
    onClick: () => void;
    onStatusChange: (next: TaskStatus) => void;
  }
>(({ t, depth, virtualStyle, match, onClick, onStatusChange }, ref) => {
  const { attributes, listeners, setNodeRef, transform, transition, isDragging } =
    useSortable({ id: t.id });
  const style: React.CSSProperties = {
    ...virtualStyle,
    ...(transform ? { transform: CSS.Transform.toString(transform) } : {}),
    ...(transition ? { transition } : {}),
  };
  if (isDragging) {
    style.transform = `${style.transform ?? ""} translateY(-1px)`;
  }
  return (
    <TaskCard
      ref={(el) => {
        setNodeRef(el);
        if (typeof ref === "function") ref(el);
        else if (ref) (ref as React.MutableRefObject<HTMLLIElement | null>).current = el;
      }}
      t={t}
      style={style}
      listeners={listeners}
      attributes={attributes}
      isDragging={isDragging}
      onClick={onClick}
      depth={depth}
      match={match}
      onStatusChange={onStatusChange}
    />
  );
});
TaskListItem.displayName = "TaskListItem";
