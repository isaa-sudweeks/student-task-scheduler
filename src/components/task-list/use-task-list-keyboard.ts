import { useState, useRef, useEffect } from "react";
import type { Virtualizer } from "@tanstack/react-virtual";

export function useTaskListKeyboard(
  count: number,
  useVirtual: boolean,
  rowVirtualizer: Virtualizer<HTMLDivElement, Element>,
  parentRef: React.RefObject<HTMLDivElement>
) {
  const [selectedIndex, setSelectedIndex] = useState(0);
  const itemRefs = useRef<(HTMLLIElement | null)[]>([]);

  useEffect(() => {
    const onKey = (e: KeyboardEvent) => {
      if (e.defaultPrevented) return;
      const target = e.target as HTMLElement | null;
      const container = parentRef.current;
      if (!container || !target || !container.contains(target)) return;
      const focusable = target.closest(
        "input, textarea, select, button, a, [contenteditable], [tabindex]:not([tabindex='-1'])"
      );
      if (focusable) return;
      const overlay = document.querySelector(
        "[role='dialog'], [role='menu'], [role='listbox']:not([data-task-list])"
      );
      if (overlay) return;
      if (e.key === "ArrowDown" || e.key.toLowerCase() === "j") {
        e.preventDefault();
        setSelectedIndex((i) => Math.min(i + 1, count - 1));
      } else if (e.key === "ArrowUp" || e.key.toLowerCase() === "k") {
        e.preventDefault();
        setSelectedIndex((i) => Math.max(i - 1, 0));
      }
    };
    window.addEventListener("keydown", onKey);
    return () => window.removeEventListener("keydown", onKey);
  }, [count, parentRef]);

  useEffect(() => {
    if (selectedIndex < 0 || selectedIndex >= count) return;
    if (useVirtual) {
      rowVirtualizer.scrollToIndex(selectedIndex);
    } else {
      itemRefs.current[selectedIndex]?.scrollIntoView({ block: "nearest" });
    }
  }, [selectedIndex, useVirtual, rowVirtualizer, count]);

  return { itemRefs };
}
