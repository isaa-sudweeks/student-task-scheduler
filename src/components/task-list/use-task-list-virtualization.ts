import { useRef, useEffect } from "react";
import { useVirtualizer } from "@tanstack/react-virtual";

export function useTaskListVirtualization(
  count: number,
  tasks: any
) {
  const parentRef = useRef<HTMLDivElement>(null);
  const rowVirtualizer = useVirtualizer({
    count,
    getScrollElement: () => parentRef.current,
    estimateSize: () => 64,
  });
  const useVirtual = count >= 20;

  useEffect(() => {
    const el = parentRef.current;
    if (!el) return;
    const handleScroll = () => {
      if (!tasks.hasNextPage || tasks.isFetchingNextPage) return;
      if (el.scrollTop + el.clientHeight >= el.scrollHeight - 100) {
        void tasks.fetchNextPage();
      }
    };
    el.addEventListener("scroll", handleScroll);
    return () => el.removeEventListener("scroll", handleScroll);
  }, [tasks]);

  return { parentRef, rowVirtualizer, useVirtual };
}
