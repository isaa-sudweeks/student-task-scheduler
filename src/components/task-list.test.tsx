// @vitest-environment jsdom
import React from 'react';
import { render, screen, cleanup, fireEvent } from '@testing-library/react';
import { describe, it, expect, vi, afterEach } from 'vitest';
import * as matchers from '@testing-library/jest-dom/matchers';
expect.extend(matchers);
import { TaskList } from './task-list';

// Capture items passed to SortableContext so we can assert compaction
const sortableItemsCalls: unknown[][] = [];
vi.mock('@dnd-kit/sortable', async () => {
  return {
    // Record each items prop received to verify it matches filtered list
    SortableContext: ({ items, children }: { items: unknown; children: React.ReactNode }) => {
      sortableItemsCalls.push(Array.isArray(items) ? [...items] : [items]);
      return <div data-testid="sortable-context">{children}</div> as any;
    },
    useSortable: () => ({
      attributes: {},
      listeners: {},
      setNodeRef: () => {},
      transform: null,
      transition: null,
    }),
    arrayMove: (arr: unknown[]) => arr,
  } as any;
});

const defaultQuery = {
  data: [
    { id: '1', title: 'Test 1', dueAt: null, status: 'DONE', subject: 'math' },
    { id: '2', title: 'Test 2', dueAt: null, status: 'TODO' },
  ],
  isLoading: false,
  error: undefined,
};
const useQueryMock = vi.fn().mockReturnValue(defaultQuery);

vi.mock('@/server/api/react', () => ({
  api: {
    useUtils: () => ({ task: { list: { invalidate: vi.fn() } } }),
    task: {
      list: {
        useQuery: (...args: unknown[]) => useQueryMock(...args),
      },
      create: {
        useMutation: () => ({
          mutate: vi.fn(),
          isPending: false,
          error: { message: 'Failed to create task' },
        }),
      },
      setDueDate: {
        useMutation: () => ({
          mutate: vi.fn(),
          isPending: false,
          error: { message: 'Failed to set due date' },
        }),
      },
      updateTitle: { useMutation: () => ({ mutate: vi.fn(), isPending: false, error: undefined }) },
      delete: { useMutation: () => ({ mutate: vi.fn(), isPending: false, error: undefined }) },
      setStatus: { useMutation: () => ({ mutate: vi.fn(), isPending: false, error: undefined }) },
      reorder: { useMutation: () => ({ mutate: vi.fn(), isPending: false, error: undefined }) },
    },
  },
}));

afterEach(() => {
  cleanup();
  useQueryMock.mockReturnValue(defaultQuery);
  sortableItemsCalls.length = 0;
});

describe('TaskList', () => {
  it('shows loading skeleton when tasks are loading', () => {
    useQueryMock.mockReturnValue({
      data: [],
      isLoading: true,
      error: undefined,
    });
    render(<TaskList />);
    expect(screen.getByLabelText('Loading tasks')).toBeInTheDocument();
  });

  it('shows error message when setting due date fails', () => {
    render(<TaskList />);
    expect(screen.getByText('Failed to set due date')).toBeInTheDocument();
  });

  it('renders subject badge', () => {
    render(<TaskList />);
    expect(screen.getByText('math')).toBeInTheDocument();
  });

  it('filters tasks based on search query', () => {
    render(<TaskList />);
    const input = screen.getByPlaceholderText('Search tasks...');
    fireEvent.change(input, { target: { value: 'Nope' } });
    expect(screen.queryByText('Test')).not.toBeInTheDocument();
    expect(screen.getByText('No tasks.')).toBeInTheDocument();
  });

  it('filters by title only and ignores subject', () => {
    useQueryMock.mockReturnValueOnce({
      data: [
        { id: '1', title: 'Read book', dueAt: null, subject: 'Math' },
        { id: '2', title: 'Math homework', dueAt: null, subject: 'English' },
      ],
      isLoading: false,
      error: undefined,
    });
    render(<TaskList />);
    const input = screen.getByPlaceholderText('Search tasks...');
    fireEvent.change(input, { target: { value: 'math' } });
    // Should match by title (subject alone should not matter)
    expect(screen.getByDisplayValue('Math homework')).toBeInTheDocument();
  });

  it('displays completed task ratio', () => {
    render(<TaskList />);
    expect(screen.getByText('1/2 completed')).toBeInTheDocument();
  });

  it('compacts visible tasks to the top when filtering', async () => {
    useQueryMock.mockReturnValue({
      data: [
        { id: 'a', title: 'Alpha', dueAt: null, status: 'TODO' },
        { id: 'b', title: 'Beta', dueAt: null, status: 'TODO' },
        { id: 'c', title: 'Gamma', dueAt: null, status: 'TODO' },
      ],
      isLoading: false,
      error: undefined,
    });
    render(<TaskList />);

    // Type a query that matches only one item
    const input = screen.getByPlaceholderText('Search tasks...');
    fireEvent.change(input, { target: { value: 'gamma' } });

    // The last SortableContext call should receive only the filtered ids
    const lastCall = sortableItemsCalls[sortableItemsCalls.length - 1] as string[];
    expect(lastCall).toEqual(['c']);
    // Only the matching task remains in the DOM (no gaps)
    expect(screen.getByDisplayValue('Gamma')).toBeInTheDocument();
    expect(screen.queryByDisplayValue('Alpha')).not.toBeInTheDocument();
    expect(screen.queryByDisplayValue('Beta')).not.toBeInTheDocument();
  });
});
