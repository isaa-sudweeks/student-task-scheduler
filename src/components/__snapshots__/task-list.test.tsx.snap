// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TaskList > applies responsive max height for long lists 1`] = `
<div>
  <div
    class="w-full space-y-3 md:w-auto"
  >
    <input
      class="w-full bg-transparent border-0 px-0 py-1 outline-none placeholder:text-muted-foreground"
      placeholder="Search tasks..."
      type="text"
      value=""
    />
    <div
      class="flex items-center justify-between"
    >
      <div
        aria-label="Task filter"
        class="flex gap-2 items-center"
        role="tablist"
      >
        <button
          aria-selected="true"
          class="bg-transparent px-3 py-1 text-sm border-b-2 transition border-black dark:border-white"
          role="tab"
          type="button"
        >
          All
        </button>
        <button
          aria-selected="false"
          class="bg-transparent px-3 py-1 text-sm border-b-2 transition border-transparent"
          role="tab"
          type="button"
        >
          Overdue
        </button>
        <button
          aria-selected="false"
          class="bg-transparent px-3 py-1 text-sm border-b-2 transition border-transparent"
          role="tab"
          type="button"
        >
          Today
        </button>
        <button
          aria-selected="false"
          class="bg-transparent px-3 py-1 text-sm border-b-2 transition border-transparent"
          role="tab"
          type="button"
        >
          Archive
        </button>
        <div
          class="relative ml-2"
        >
          <svg
            class="lucide lucide-tag pointer-events-none absolute left-2 top-1/2 h-4 w-4 -translate-y-1/2 text-slate-500"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12.586 2.586A2 2 0 0 0 11.172 2H4a2 2 0 0 0-2 2v7.172a2 2 0 0 0 .586 1.414l8.704 8.704a2.426 2.426 0 0 0 3.42 0l6.58-6.58a2.426 2.426 0 0 0 0-3.42z"
            />
            <circle
              cx="7.5"
              cy="7.5"
              fill="currentColor"
              r=".5"
            />
          </svg>
          <select
            aria-label="Subject filter"
            class="appearance-none rounded-full border border-slate-200 bg-slate-100 py-1.5 pl-8 pr-8 text-sm text-slate-800 shadow-sm transition hover:bg-slate-50 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-indigo-500 focus-visible:ring-offset-2 dark:border-slate-700 dark:bg-slate-800/60 dark:text-slate-200 dark:hover:bg-slate-800"
            title="Filter by subject"
          >
            <option
              value=""
            >
              All subjects
            </option>
          </select>
          <svg
            class="lucide lucide-chevron-down pointer-events-none absolute right-2 top-1/2 h-4 w-4 -translate-y-1/2 text-slate-500"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m6 9 6 6 6-6"
            />
          </svg>
        </div>
        <div
          class="relative ml-2"
        >
          <svg
            class="lucide lucide-flag pointer-events-none absolute left-2 top-1/2 h-4 w-4 -translate-y-1/2 text-slate-500"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M4 15s1-1 4-1 5 2 8 2 4-1 4-1V3s-1 1-4 1-5-2-8-2-4 1-4 1z"
            />
            <line
              x1="4"
              x2="4"
              y1="22"
              y2="15"
            />
          </svg>
          <select
            aria-label="Priority filter"
            class="appearance-none rounded-full border border-slate-200 bg-slate-100 py-1.5 pl-8 pr-8 text-sm text-slate-800 shadow-sm transition hover:bg-slate-50 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-indigo-500 focus-visible:ring-offset-2 dark:border-slate-700 dark:bg-slate-800/60 dark:text-slate-200 dark:hover:bg-slate-800"
            title="Filter by priority"
          >
            <option
              value=""
            >
              All priorities
            </option>
            <option
              value="HIGH"
            >
              High
            </option>
            <option
              value="MEDIUM"
            >
              Medium
            </option>
            <option
              value="LOW"
            >
              Low
            </option>
          </select>
          <svg
            class="lucide lucide-chevron-down pointer-events-none absolute right-2 top-1/2 h-4 w-4 -translate-y-1/2 text-slate-500"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m6 9 6 6 6-6"
            />
          </svg>
        </div>
      </div>
      <p
        class="text-xs text-gray-500 dark:text-gray-400"
      >
        0
         archived
      </p>
    </div>
    <div>
      <div
        data-testid="sortable-context"
      >
        <div
          class="overflow-auto max-h-[50vh] md:max-h-[600px]"
          data-testid="task-scroll"
        >
          <div
            style="height: 0px; position: relative;"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
